1 identify and review analytical sample

crispy-wonton opened the pull request:
---
Addresses #1 

# Description
In this PR, I have added code to:
- Add package requirements
- Add notebook to create analytical sample based on exclusion criteria, review proportions of NaN values for key variables, and review proportions of data lost when removing rows
- NB a separate preprocessing.py has been created to generate and upload the analytical sample to S3 for use in a final script to produce the final output. This will appear in a separate PR to keep PRs smaller.

This branch will be merged when changes made and approved.

# Instructions for Reviewer
To run:
1. Clone repo and follow instructions in README to set up conda env
2. `pip install jupytext`
3. `jupytext --to notebook asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py`

Please could you:
- Review code for errors
- Check exclusion criteria are sufficient and correctly represented
- Let me know your thoughts on the cost distribution in removed vs retained rows. Are you concerned that a disproportionate amount of higher costs have been removed?
- Let me know if you want me to explore anything further. Note I can add code to review if any particular property archetypes are disproportionately removed when the archetype code has been added and reviewed.

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
add requirements

crispy-wonton committed:
add code to identify and review analytical sample
based on agreed exclusion criteria

crispy-wonton committed:
convert 'unknown' string values to NaN

Commit:
Code review additions

Commit:
Review comments for PR #1

danlewis85 commented:
This is generally fine and I'm happy with the work.

However, there are a couple of coding points that could make your code clearer/easier in future.

Pandas DataFrames and Series methods are mostly vectorised, so you don't need to 'apply()' them. This is notable with any which is already a vectorised method.
Applying a lambda that returns a boolean can probably be achieved with logic statements in many cases, so do check.
Boolean Series do not need to be tested for equivalence (e.g. df['boolean_series'] == True) as they already act as filters, if required they can be inverted with tilde (e.g. df['boolean_series']).
In terms of the consideration of missing data, I've unpacked the problem at the bottom of the notebook and added some commentary.

Check you're happy with the proposed changes to code, update and merge.

danlewis85 commented:
Oh, and with flat files I think there is value in specifying a data schema - check what I've done and make sure you're happy.

danlewis85 reviewed (changes requested):
As per comments in conversation.

1. Tidy up code for clarity/simplicity.
2. Review and add data schema as appropriate (could go in config?)
3. Consider brief analysis of missing added to bottom of notebook, let me know if it doesn't make sense.

Once done, happy to merge.

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+# %%
+# I'm keen we load flat files according to a known schema
+data_schema = {
+    "original_mcs_index": "int32",
+    "version": "int8",
```
The schema is great, thank you. I have a few questions:
1. How did you generate the schema?
2. Did you manually determine the int size for int columns? If so, how?
3. Why did you specify int bit size but not float bit size?
4. What's the purpose of setting dtype to `str` in pandas if pandas stores string columns as `object`? Does it actually store info that it's a string?

danlewis85 replied:
1. There's a general schema here: https://docs.google.com/spreadsheets/d/1UE8mMPeIeDIE_Rm7UmR9gWkGGJEUhCyjnSH2e22pfMg/edit#gid=1273932954
2. Yes, by looking at the max and min values. 
3. I don't understand the implications of floating point size well enough. I suspect that float32 would be adequate in all cases though.
4. Setting `str` to string avoids mixed type objects - in this case not setting `str` creates columns with strings and floating point type null objects.

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+#       format_version: '1.3'
+#       jupytext_version: 1.16.1
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
```
Did you manually change the display_name or did it auto-update with your changes to the file?

danlewis85 replied:
Auto update.

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+
+# %%
+# filter for just ASHP in dropped rows.
+dropped_rows = dropped_rows.loc[lambda df: df["tech_type"] == "Air Source Heat Pump", :]
```
Good point

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+plt.show()
+
+# %% [markdown]
+# ## Note on comparison
```
Thanks for adding this section and the code below, it's great.

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+#
+# Descriptively the dropped rows and sample look similar - similar means, dropped rows has a higher std. deviation, which is interesting, sample again a bit higher at each quartile.
+#
+# Let's see what this actually looks like graphed out. NB I'm going to truncate the cost variable at £30,000 which is roughly the 99th percentile of the distribution, as I'm not actually that interested in the tail.
```
Which dataset did you determine the 99th percentile from and how, for reproducibility?

danlewis85 replied:
I don't think this really matters, it's just an indicative value. In reality you could choose any number that meaningfully removes the tail whilst retaining the vast majority of the data points.

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+sample["cost"].describe()
+
+# %%
+f, ax = plt.subplots(figsize=(8, 5))
+
+ax.hist(
+    dropped_rows.loc[lambda df: df["cost"] < 30_000, "cost"],
+    density=True,
+    histtype="bar",
+    alpha=0.5,
+    bins=20,
+    color="indianred",
+    label="Dropped",
```
``
```suggestion
binned_data, bin_defs = pd.cut(sample[sample['cost'] < 30000]['cost'], bins=20, retbins=True)
f, ax = plt.subplots(figsize=(8, 5))

ax.hist(
    dropped_rows.loc[lambda df: df["cost"] < 30_000, "cost"],
    density=True,
    histtype="bar",
    alpha=0.5,
    bins=bin_defs,
    color="indianred",
    label="Dropped",
```
Same comment for my histogram above. Is it better to use the same bin ranges for each dataset for better comparison? It looks like it has negligible effect here because the data is truncated but is noticeable on non-truncated data.

danlewis85 replied:
Ideally you should apply the same bin ranges. I'm comfortable here that it doesn't really matter.

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+    density=True,
+    histtype="bar",
+    alpha=0.5,
+    bins=20,
```
```suggestion
    bins=bin_defs,
```

danlewis85 replied:
Nice.

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+# In this case we'll normalise the data so they're comparable.
+
+fig, ax = plt.subplots(nrows=1, ncols=2, sharey=True, figsize=(10, 3))
+ax[0].hist(sample["cost"], density=True)
```
I originally left the `density` arg out as I believe it makes the area under the graph integrate to 1 so the y values became very small, which ultimately meant I didn't have a good intuition of how to interpret the y-axis and wasn't sure about its readability in general. 

I understand the reasoning to normalise the data so they are comparable across the two samples and I guess the important thing is to compare the shapes of the distributions so the y-axis is less relevant (?). Can you input on how to interpret the y-axis?

danlewis85 replied:
The y axis is the density as you've stated. It's the absolute value relative to the sum of absolute values for the distribution. You can think of it as a scaling. It has no real interpretation, it just makes distributions with different volumes comparable while maintaining their shape.

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+# %%
+## Keep rows where drop_row is False, i.e. remove rows based on exclusion criteria
+# sample = df[df["drop_row"] == False]
+sample = df[~df["drop_row"]]
```
Sometimes i find the original version with explicit reference easier to read and understand. Is this bad practice or just a matter of preference?

danlewis85 replied:
While it is subjective, the python style guide suggests it is to be avoided: https://peps.python.org/pep-0008/#programming-recommendations:~:text=Don%E2%80%99t%20compare%20boolean,%3D%3D%20True%3A

I personally don't like it. However, this is ultimately your code. If you want to keep it that's fine. The code works so I'm not going to push for it to change.

crispy-wonton replied:
Cool I dont feel strongly here so i'll keep it the recommended way. Thanks!

crispy-wonton commented on asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py:
```diff
@@ -0,0 +1,338 @@
+property_archetype_cols = ["CONSTRUCTION_AGE_BAND", "BUILT_FORM", "PROPERTY_TYPE"]
+# df["property_archetype_data_na"] = df[property_archetype_cols].isna().apply(any, axis=1)
+# No need to apply
+df["property_archetype_data_na"] = df[property_archetype_cols].isna().any(axis=1)
```
thanks, will use this format from now on

crispy-wonton reviewed (commented):
Thanks for your review and additions @danlewis85 , it's very helpful. I've left some questions to improve my understanding and then I'll commit the clean up and updates.

crispy-wonton committed:
add json data schema for importing joined MCS-EPC flat file

crispy-wonton committed:
make changes requested in PR review:
- clean up code
- use imported data schema for flat files
- restructure analysis for clarity

crispy-wonton commented:
I have made the following changes based on the PR review:

- Add json file with data schema (and read json function) which can be imported to scripts & notebooks
- Cleaned up notebook to simplify code
- Replaced my original histogram plots with yours as yours is better and I don't think mine added anything else
- I kept your notes and explanation as I think it's a useful summary and interpretation for reference.

Let me know if you're happy to merge :)

crispy-wonton committed:
change object to string type in json schema

danlewis85 reviewed (approved):
Happy to merge.
